Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt

```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
 
 import com.scribd.armadillo.actions.Action
 
-sealed class ArmadilloException(exception: Exception) : Exception(exception) {
+sealed class ArmadilloException(cause: Throwable? = null,
+                                message: String = "Unhandled Armadillo Error")
+    : Exception(message, cause) {
     abstract val errorCode: Int
 }
 
 /**
  * Errors in Armadillo
  */
-data class EngineNotInitialized(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class EngineNotInitialized(message: String)
+    : ArmadilloException(message = "Armadillo Engine is not initialized: $message") {
     override val errorCode = 100
 }
 
-object TransportControlsNull : ArmadilloException(Exception()) {
+class TransportControlsNull
+    : ArmadilloException(message = "Internal error, the transport controls are null. Are the controls being initialized?") {
     override val errorCode = 101
 }
 
-data class MissingDataException(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class MissingDataException(message: String)
+    : ArmadilloException(message = "Internal error, unexpectedly missing internal data: $message") {
     override val errorCode = 102
 }
 
-data class ActionBeforeSetup(val action: Action) : ArmadilloException(exception = Exception(action.name)) {
+data class ActionBeforeSetup(val action: Action)
+    : ArmadilloException(message = "Internal Error, using an Action before setup is finished: ${action.name}") {
     override val errorCode = 103
 }
 
-data class UnrecognizedAction(val action: Action) : ArmadilloException(exception = Exception(action.name)) {
+data class UnrecognizedAction(val action: Action)
+    : ArmadilloException(message = "Internal error, unknown action being taken: ${action.name}") {
     override val errorCode = 104
 }
 
-object NoPlaybackInfo : ArmadilloException(Exception()) {
+class NoPlaybackInfo : ArmadilloException(message = "Internal error, playback info is missing. Is the player initialized?") {
     override val errorCode = 105
 }
 
-object InvalidPlaybackState : ArmadilloException(Exception()) {
+class InvalidPlaybackState
+    : ArmadilloException(message = "Internal Error, playback state is missing. Has the player been initialized or destroyed?") {
     override val errorCode = 106
 }
 
-data class InvalidRequest(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class InvalidRequest(message: String)
```

Comment:
why not keep this as reason and still use `data class`?

Reply:
There is no need to use `reason` when `message` already exists in the superclass and using `reason` is hiding users from finding `message`. People are expecting `message` and not any weird custom attribute.

Reply:
I'm not suggesting we remove `message` from super class. Just suggesting we use `reason` for the child constructor and construct message using reason .. something like you did with `EngineNotInitialized`
```

data class InvalidRequest(val reason: String) : ArmadilloException(exception = Exception("Armadillo found invalid request: $reason"))